miniapps/electromagnetics/maxwell_solver.cpp

```diff
@@ -0,0 +1,673 @@
+{
+   if ( pcg_.find(idt) == pcg_.end() )
+   {
+      if ( myid_ == 0 && logging_ > 0 )
+      {
+         cout << "Creating implicit operator for dt = " << dt << endl;
+      }
+
+      a1_[idt] = new ParBilinearForm(HCurlFESpace_);
+      a1_[idt]->AddDomainIntegrator(
+         new VectorFEMassIntegrator(epsCoef_));
+
+      if ( idt != 0 )
+      {
+         dtCoef_[idt] = new ConstantCoefficient(0.5 * dt);
+         if ( sigmaCoef_ )
+         {
+            dtSigmaCoef_[idt] = new TransformedCoefficient(dtCoef_[idt],
+                                                           sigmaCoef_,
+                                                           prodFunc);
+            a1_[idt]->AddDomainIntegrator(
+               new VectorFEMassIntegrator(dtSigmaCoef_[idt]));
+         }
+         if ( etaInvCoef_ )
+         {
+            dtEtaInvCoef_[idt] = new TransformedCoefficient(dtCoef_[idt],
+                                                            etaInvCoef_,
+                                                            prodFunc);
+            a1_[idt]->AddBoundaryIntegrator(
+               new VectorFEMassIntegrator(dtEtaInvCoef_[idt]),
+               const_cast<Array<int>&>(abc_marker_));
+         }
+      }
+
+      a1_[idt]->Assemble();
+      a1_[idt]->Finalize();
+      A1_[idt] = a1_[idt]->ParallelAssemble();
+
+      diagScale_[idt] = new HypreDiagScale(*A1_[idt]);
+      pcg_[idt] = new HyprePCG(*A1_[idt]);
+      pcg_[idt]->SetTol(1.0e-12);
+      pcg_[idt]->SetMaxIter(200);
+      pcg_[idt]->SetPrintLevel(0);
+      pcg_[idt]->SetPreconditioner(*diagScale_[idt]);
+   }
+}
+
+void
+MaxwellSolver::implicitSolve(double dt, const Vector &B, Vector &dEdt) const
+{
+   int idt = hCurlLosses_ ? ((int)(dtScale_ * dt / dtMax_)) : 0;
+
+   b_->Distribute(B);
+   weakCurlMuInv_->Mult(*b_, *rhs_);
+
+   if ( hCurlLosses_ )
+   {
+      e_->Distribute(*E_);
+      hCurlLosses_->AddMult(*e_, *rhs_, -1.0);
+   }
+
+   if ( jd_ )
+   {
+      jCoef_->SetTime(t); // 't' is member data from mfem::TimeDependentOperator
+      jd_->Assemble();
+      *rhs_ -= *jd_;
+   }
+
+   if ( dEdtBCCoef_ )
+   {
+      dedt_->ProjectBdrCoefficientTangent(*dEdtBCCoef_,
+                                          const_cast<Array<int>&>(dbc_marker_));
+   }
+
+   // Create objects and matrices for solving with the given time step
+   setupSolver(idt, dt);
+
+   // Apply essential BCs and determine true DoFs for the right hand side
+   a1_[idt]->FormLinearSystem(dbc_dofs_, *dedt_, *rhs_, *A1_[idt], dEdt, *RHS_);
+
+   // Solve for the time derivative of the electric field (true DoFs)
+   pcg_[idt]->Mult(*RHS_, dEdt);
+
+   // Distribute shared DoFs to relevant processors
+   a1_[idt]->RecoverFEMSolution(dEdt, *rhs_, *dedt_);
+}
+
+void
+MaxwellSolver::SyncGridFuncs()
+{
+   e_->Distribute(*E_);
+   b_->Distribute(*B_);
+}
+
+double
+MaxwellSolver::GetMaximumTimeStep() const
+{
+   if ( dtMax_ > 0.0 )
+   {
+      return dtMax_;
```

Comment:
Are there any time dependent coefficients in the update operator that can affect the eigenvalues?

Reply:
No, the material coefficients are assumed to be constant in time.